provider/keycloak_openid_user_attribute_protocol_mapper_test.go

```diff
@@ -0,0 +1,329 @@
+
+	resource.Test(t, resource.TestCase{
+		Providers:    testAccProviders,
+		PreCheck:     func() { testAccPreCheck(t) },
+		CheckDestroy: testAccKeycloakOpenIdUserAttributeProtocolMapperDestroy(),
+		Steps: []resource.TestStep{
+			{
+
+				Config: testKeycloakOpenIdUserAttributeProtocolMapper_basic_clientScope(realmName, clientScopeId, mapperName),
+				Check:  testKeycloakOpenIdUserAttributeProtocolMapperExists(resourceName),
+			},
+		},
+	})
+}
+
+func TestAccKeycloakOpenIdUserAttributeProtocolMapper_update(t *testing.T) {
+	realmName := "terraform-realm-" + acctest.RandString(10)
+	clientId := "terraform-client-" + acctest.RandString(10)
+	mapperName := "terraform-openid-connect-user-attribute-mapper-" + acctest.RandString(5)
+
+	attributeName := "claim-" + acctest.RandString(10)
+	updatedAttributeName := "claim-update-" + acctest.RandString(10)
+	resourceName := "keycloak_openid_user_attribute_protocol_mapper.user-attribute-mapper"
+
+	resource.Test(t, resource.TestCase{
+		Providers:    testAccProviders,
+		PreCheck:     func() { testAccPreCheck(t) },
+		CheckDestroy: testAccKeycloakOpenIdUserAttributeProtocolMapperDestroy(),
+		Steps: []resource.TestStep{
+			{
+				Config: testKeycloakOpenIdUserAttributeProtocolMapper_claim(realmName, clientId, mapperName, attributeName),
+				Check:  testKeycloakOpenIdUserAttributeProtocolMapperExists(resourceName),
+			},
+			{
+				Config: testKeycloakOpenIdUserAttributeProtocolMapper_claim(realmName, clientId, mapperName, updatedAttributeName),
+				Check:  testKeycloakOpenIdUserAttributeProtocolMapperExists(resourceName),
+			},
+		},
+	})
+}
+
+func TestAccKeycloakOpenIdUserAttributeProtocolMapper_validateClientOrClientScopeSet(t *testing.T) {
+	realmName := "terraform-realm-" + acctest.RandString(10)
+	mapperName := "terraform-openid-connect-user-attribute-mapper-" + acctest.RandString(10)
+
+	resource.Test(t, resource.TestCase{
+		Providers:    testAccProviders,
+		PreCheck:     func() { testAccPreCheck(t) },
+		CheckDestroy: testAccKeycloakOpenIdUserAttributeProtocolMapperDestroy(),
+		Steps: []resource.TestStep{
+			{
+				Config:      testKeycloakOpenIdUserAttributeProtocolMapper_validation(realmName, mapperName),
+				ExpectError: regexp.MustCompile("one of client_id or client_scope_id must be set"),
+			},
+		},
+	})
+}
+
+func TestAccKeycloakOpenIdUserAttributeProtocolMapper_validateClaimValueType(t *testing.T) {
+	realmName := "terraform-realm-" + acctest.RandString(10)
+	mapperName := "terraform-openid-connect-user-attribute-mapper-" + acctest.RandString(10)
+	invalidClaimValueType := acctest.RandString(5)
+	config := fmt.Sprintf(`
+resource "keycloak_realm" "realm" {
+	realm = "%s"
+}
+
+resource "keycloak_openid_user_attribute_protocol_mapper" "user-attribute-mapper-validation" {
+  	name = "%s"
+	realm_id = "${keycloak_realm.realm.id}"
+  	user_attribute = "foo"
+  	claim_name = "bar"
+	claim_value_type = "%s"
+}
+`, realmName, mapperName, invalidClaimValueType)
```

Comment:
I don't have a good reason for this, but conventionally this is pulled out into a function that lives near the bottom of this file.